pkg/compact/compact.go

```diff
@@ -337,6 +353,105 @@ func (c *Syncer) GarbageCollect(ctx context.Context) error {
 	return nil
 }
 
+// IndexCache creates index cache file for blocks from the bucket if blocks don't have index cache.
+func (c *Syncer) IndexCache(ctx context.Context, compactDir string) error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	indexCacheDir := filepath.Join(compactDir, "index_cache")
+	if err := os.MkdirAll(indexCacheDir, 0777); err != nil {
+		return errors.Wrap(err, "create index cache dir")
+	}
+	defer func() {
+		if err := os.RemoveAll(indexCacheDir); err != nil {
+			level.Error(c.logger).Log("msg", "failed to remove temp index cache dir", "path", indexCacheDir, "err", err)
+		}
+	}()
+
+	level.Info(c.logger).Log("msg", "start index cache processing")
+
+	for id, meta := range c.blocks {
+		err := c.cacheIndex(ctx, indexCacheDir, id, meta)
+		if err != nil {
+			c.metrics.indexCacheFailures.Inc()
+			return err
+		}
+	}
+
+	c.metrics.indexCacheTraverse.Inc()
+	return nil
+}
+
+func (c *Syncer) cacheIndex(ctx context.Context, indexCacheDir string, id ulid.ULID, meta *metadata.Meta) error {
+	// New version of compactor pushes index cache along with data block.
+	// Skip blocks with existing cache or from sidecar.
+	if meta.Thanos.Source == metadata.SidecarSource || meta.Thanos.Cache != 0 {
+		return nil
+	}
+
+	bdir := filepath.Join(indexCacheDir, id.String())
+	if err := os.MkdirAll(bdir, 0777); err != nil {
+		return errors.Wrap(err, "create block dir")
+	}
+
+	cachePath := filepath.Join(bdir, block.IndexCacheFilename)
+	cache := path.Join(id.String(), block.IndexCacheFilename)
+	metaFile := filepath.Join(bdir, metadata.MetaFilename)
+
+	updateMeta := func() error {
+		var newMeta *metadata.Meta
+		*newMeta = *meta
+		newMeta.Thanos.Cache = metadata.IndexCacheJson
+		if err := metadata.Write(c.logger, bdir, newMeta); err != nil {
+			return errors.Wrap(err, "write new meta")
+		}
+
+		// TODO: override meta? error on upload?
+		if err := objstore.UploadFile(ctx, c.logger, c.bkt, metaFile, metadata.MetaFilename); err != nil {
```

Comment:
Yea, just don't do it (: Why we are adding this `meta.Thanos.Cache`? version make sense but it should part of index cache actually.